File: heapsort/heapsort.cpp
#include <iostream>
#include <array>
#include <chrono>
#include <iterator>
#include <numeric>
#include <vector>
#include <algorithm>

void swap(std::array<int, 100'000>& arr, int *g, int *b) {
    int temp;
    temp = arr[*g];
    arr[*g] = arr[*b];
    arr[*b] = temp;
}

void maxHeapify(std::array<int, 100'000>& arr, int parent, int last) {
    int child;
    child = 2 * parent;
    while (child <= last) {
        if ((child + 1 <= last) && (arr[child + 1] > arr[child]))
            child++;
        if (arr[parent] < arr[child])
            swap(arr, &parent, &child);
        parent = child;
        child = 2 * parent;
    }
}

void buildHeap(std::array<int, 100'000>& arr, int last) {
    int i;
    for (i = last / 2; i > 0; i--) {
        maxHeapify(arr, i, last);
    }
}

void heapSort(std::array<int, 100'000>& arr, int last) {
    int l;
    l = 1;
    buildHeap(arr, last);
    while (last > 1) {
        swap(arr, &l, &last);
        last--;
        maxHeapify(arr, 1, last);
    }
}

using namespace std::chrono;
int main()
{
    std::vector<double> times;
    std::array<int, 100'000> nums;

    for (int i = 0; i < 1'000; i++)
    {
        std::generate(nums.begin(), nums.end(), &rand);
        milliseconds now = duration_cast<milliseconds>(system_clock::now().time_since_epoch());
        heapSort(nums, 100'000);
        milliseconds later = duration_cast<milliseconds>(system_clock::now().time_since_epoch());
        times.push_back(duration_cast<milliseconds>(later - now).count());
    }
    std::cout << std::accumulate(times.begin(), times.end(), 0.0) / times.size() << "\n";

    return 0;
}

File: heapsort/heapsort.js
function sort( arr,N)
    { 
        for (var i = Math.floor(N / 2) - 1; i >= 0; i--)
            heapify(arr, N, i);
 
        for (var i = N - 1; i > 0; i--) {
            var temp = arr[0];
            arr[0] = arr[i];
            arr[i] = temp;
 
            heapify(arr, i, 0);
        }
    }
 
    function heapify(arr, N, i)
    {
        var largest = i;
        var l = 2 * i + 1; 
        var r = 2 * i + 2; 
 
        if (l < N && arr[l] > arr[largest])
            largest = l;
 
        if (r < N && arr[r] > arr[largest])
            largest = r;
 
        if (largest != i) {
            var swap = arr[i];
            arr[i] = arr[largest];
            arr[largest] = swap;
 
            heapify(arr, N, largest);
        }
    }

times = []
for (i = 0; i < 1000; ++i) {
	nums = Array.from({length: 100000}, () => Math.floor(Math.random() * 100000));
	now = Date.now();
    sort(nums, 100000 - 1);
	later = Date.now();
	times.push(later - now);
}
console.log(times.reduce((a,b) => a+b) / times.length)
